h/authorization.py

```diff
@@ -113,15 +175,48 @@ def save_bearer_token(self, token, request):
     def validate_bearer_token(self, token, scopes, request):
         if token is None:
             return False
-        client = request.registry.web_client
-        ttl = auth.DEFAULT_TTL
+
         try:
-            token = auth.decode_token(token, client.client_secret, ttl)
-        except auth.TokenInvalid:
+            decoded = jwt.decode(token, verify=False)
+        except jwt.DecodeError:
             return False
-        request.client = client  # TODO: 3rd party authorizations
-        request.user = token.get('userId')
-        request.scopes = token.get('scopes', [])
+
+        normalize_authorization_token(request, decoded)  # bw compat
+
+        if any(key not in decoded for key in ['aud', 'exp', 'iss', 'sub']):
+            return False
+
+        t = posix_seconds(now())
+        if decoded['exp'] <= t:
+            return False
+        if 'nbf' in decoded and decoded['nbf'] > t:
+            return False
+
+        user = decoded['sub']
+        if not isinstance(user, string_types):
+            return False
+
+        aud = decoded['aud']
+        if aud != request.host_url:
+            return False
+
+        iss = decoded['iss']
+        if is_web_client(request, iss):
+            client = request.registry.web_client
+            scopes = WEB_SCOPES
+        else:
+            # TODO: 3rd party authorizations
+            return False
+
+        try:
+            # TODO: client certificates instead of secrets
+            jwt.decode(token, client.client_secret)
+        except jwt.DecodeError:
+            return False
```

Comment:
This is now a seriously long method, how about we break it up into smaller functions that do each check.